File: plugins/reverb/spring-reverb/SpringReverb.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Spring Reverb (native port) — 8-stage allpass dispersion feeding two
// cross-coupled delay lines (the springs). Drippy, chirpy boing of a
// Fender guitar amp tank or Hammond organ.
public final class SpringReverb implements VocalMonitorNativePlugin {
    private static final int[] AP_LENS = { 37, 53, 71, 89, 107, 131, 149, 167 };
    private float[][] aps;
    private int[] apIdx;
    private float[] d1, d2;
    private int d1Len, d2Len;
    private int d1Idx = 0, d2Idx = 0;
    private float lp1 = 0f, lp2 = 0f;
    private float hp1Prev = 0f, hp1Out = 0f;
    private float decay = 0.65f, boing = 0.7f, tone = 0.45f, mix = 0.4f;

    @Override
    public void init(int sr) {
        float srScale = sr / 44100f;
        aps = new float[AP_LENS.length][];
        apIdx = new int[AP_LENS.length];
        for (int a = 0; a < AP_LENS.length; a++) {
            int L = (int) Math.ceil(AP_LENS[a] * srScale);
            aps[a] = new float[L];
            apIdx[a] = 0;
        }
        d1Len = (int) Math.ceil(sr * 0.13);
        d2Len = (int) Math.ceil(sr * 0.16);
        d1 = new float[d1Len];
        d2 = new float[d2Len];
        d1Idx = d2Idx = 0;
        lp1 = lp2 = 0f;
        hp1Prev = hp1Out = 0f;
    }

    @Override public String[] parameterNames() { return new String[] { "decay", "boing", "tone", "mix" }; }
    @Override public float parameterMin(String n) { return 0f; }
    @Override public float parameterMax(String n) {
        return "decay".equals(n) ? 0.92f : 1f;
    }
    @Override public float parameterDefault(String n) {
        switch (n) {
            case "decay": return 0.65f;
            case "boing": return 0.7f;
            case "tone":  return 0.45f;
            case "mix":   return 0.4f;
            default:      return 0f;
        }
    }
    @Override public String parameterLabel(String n) {
        switch (n) {
            case "decay": return "Decay";
            case "boing": return "Boing";
            case "tone":  return "Tone";
            case "mix":   return "Mix";
            default:      return n;
        }
    }
    @Override public void setParameter(String n, float v) {
        switch (n) {
            case "decay": decay = v; break;
            case "boing": boing = v; break;
            case "tone":  tone = v; break;
            case "mix":   mix = v; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        final float apG = 0.3f + 0.5f * boing;
        final float lpA = 0.05f + 0.85f * tone;
        final float hpA = 0.99f;
        final float decayLocal = decay;
        final float mixLocal = mix;
        final float dry = 1f - mixLocal;
        final int numAps = AP_LENS.length;
        final float[][] apsLocal = aps;
        final int[] apIdxLocal = apIdx;
        final float[] d1Local = d1, d2Local = d2;
        final int d1L = d1Len, d2L = d2Len;
        int d1i = d1Idx, d2i = d2Idx;
        float l1 = lp1, l2 = lp2;
        float hpP = hp1Prev, hpO = hp1Out;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            float x = input[i];
            hpO = hpA * (hpO + x - hpP);
            hpP = x;
            float hp = hpO;
            int d1r = d1i - (d1L - 1); if (d1r < 0) d1r += d1L;
            int d2r = d2i - (d2L - 1); if (d2r < 0) d2r += d2L;
            float s1Out = d1Local[d1r];
            float s2Out = d2Local[d2r];
            float v = hp + s1Out * decayLocal * 0.7f + s2Out * decayLocal * 0.3f;
            for (int a = 0; a < numAps; a++) {
                float[] ab = apsLocal[a];
                int L = ab.length;
                int idx = apIdxLocal[a];
                int r = idx - (L - 1); if (r < 0) r += L;
                float del = ab[r];
                float inAp = v + del * apG;
                ab[idx] = inAp;
                v = del - inAp * apG;
                idx++; if (idx >= L) idx = 0;
                apIdxLocal[a] = idx;
            }
            l1 = l1 + lpA * (v - l1);
            l2 = l2 + lpA * (v - l2);
            d1Local[d1i] = l1 + s2Out * decayLocal * 0.2f;
            d2Local[d2i] = l2 + s1Out * decayLocal * 0.2f;
            d1i++; if (d1i >= d1L) d1i = 0;
            d2i++; if (d2i >= d2L) d2i = 0;
            float wet = (s1Out + s2Out) * 0.5f;
            output[i] = x * dry + wet * mixLocal;
        }
        d1Idx = d1i; d2Idx = d2i;
        lp1 = l1; lp2 = l2;
        hp1Prev = hpP; hp1Out = hpO;
    }
}

File: plugins/reverb/spring-reverb/spring-reverb.js
// Spring Reverb — recreates the dispersive, twangy character of a real
// spring tank by cascading a chain of allpass sections (which delay
// different frequencies by different amounts — that's "dispersion", the
// thing that gives a struck spring its chirpy "boing" before the tail
// settles). Two short delay lines with cross-feedback simulate the two
// springs of a classic Fender / Hammond tank. Add some highpass on the
// way in and lowpass in the loop so the tail sounds metallic-but-warm,
// the way an old guitar amp's tank does.

function SpringReverbProcessor() {
  var srScale = sampleRate / 44100;
  // Allpass dispersion chain — 8 sections with prime-ish lengths so the
  // group-delay curve isn't periodic.
  this.apLens = [37, 53, 71, 89, 107, 131, 149, 167];
  this.aps = [];
  this.apIdx = [];
  for (var a = 0; a < this.apLens.length; a++) {
    var L = Math.ceil(this.apLens[a] * srScale);
    var b = new Array(L);
    for (var k = 0; k < L; k++) b[k] = 0;
    this.aps.push(b);
    this.apIdx.push(0);
  }
  // Two parallel delay lines — the "springs". Long enough for ~120 ms.
  var d1Len = Math.ceil(sampleRate * 0.13);
  var d2Len = Math.ceil(sampleRate * 0.16);
  this.d1 = new Array(d1Len); this.d2 = new Array(d2Len);
  for (var k = 0; k < d1Len; k++) this.d1[k] = 0;
  for (var k = 0; k < d2Len; k++) this.d2[k] = 0;
  this.d1Len = d1Len; this.d2Len = d2Len;
  this.d1Idx = 0; this.d2Idx = 0;
  this.lp1 = 0; this.lp2 = 0;
  this.hp1Prev = 0; this.hp1Out = 0;
}

SpringReverbProcessor.parameterDescriptors = [
  { name: 'decay',  label: 'Decay',  defaultValue: 0.65, minValue: 0,   maxValue: 0.92 },
  { name: 'boing',  label: 'Boing',  defaultValue: 0.7,  minValue: 0,   maxValue: 1 },
  { name: 'tone',   label: 'Tone',   defaultValue: 0.45, minValue: 0,   maxValue: 1 },
  { name: 'mix',    label: 'Mix',    defaultValue: 0.4,  minValue: 0,   maxValue: 1 }
];

SpringReverbProcessor.prototype.process = function (inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var decay = parameters.decay[0];
  var boing = parameters.boing[0];
  var tone = parameters.tone[0];
  var mix = parameters.mix[0];

  // Allpass coefficient — higher = more dispersion = more boing.
  var apG = 0.3 + 0.5 * boing;
  // Lowpass in loop (tone knob): closed = dark, open = bright.
  var lpA = 0.05 + 0.85 * tone;
  // Pre-emphasis highpass to keep the rumble out of the springs.
  var hpA = 0.99;

  var numAps = this.apLens.length;

  for (var i = 0; i < input.length; i++) {
    var x = input[i];
    // Highpass the input.
    this.hp1Out = hpA * (this.hp1Out + x - this.hp1Prev);
    this.hp1Prev = x;
    var hp = this.hp1Out;

    // Cross-feedback from both springs into the input of the AP chain.
    var d1R = this.d1Idx - (this.d1Len - 1); if (d1R < 0) d1R += this.d1Len;
    var d2R = this.d2Idx - (this.d2Len - 1); if (d2R < 0) d2R += this.d2Len;
    var s1Out = this.d1[d1R];
    var s2Out = this.d2[d2R];

    var v = hp + s1Out * decay * 0.7 + s2Out * decay * 0.3;

    // Run through allpass dispersion chain.
    for (var a = 0; a < numAps; a++) {
      var L = this.aps[a].length;
      var idx = this.apIdx[a];
      var r = idx - (L - 1); if (r < 0) r += L;
      var del = this.aps[a][r];
      var inAp = v + del * apG;
      this.aps[a][idx] = inAp;
      v = del - inAp * apG;
      idx++; if (idx >= L) idx = 0;
      this.apIdx[a] = idx;
    }

    // Damping LP, then write into both springs (cross-coupled).
    this.lp1 = this.lp1 + lpA * (v - this.lp1);
    this.lp2 = this.lp2 + lpA * (v - this.lp2);
    this.d1[this.d1Idx] = this.lp1 + s2Out * decay * 0.2;
    this.d2[this.d2Idx] = this.lp2 + s1Out * decay * 0.2;
    this.d1Idx++; if (this.d1Idx >= this.d1Len) this.d1Idx = 0;
    this.d2Idx++; if (this.d2Idx >= this.d2Len) this.d2Idx = 0;

    var wet = (s1Out + s2Out) * 0.5;
    output[i] = x * (1 - mix) + wet * mix;
  }
  return true;
};

registerProcessor('spring-reverb', SpringReverbProcessor);
